improve: send device id in all RAT events (SDKCF-6625)

Esakkiraja-Pothikannan opened the pull request:
# Description
- Send `device_id` as a custom parameter in all RAT events

## Links
SDKCF-6625

# Checklist
- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md)
- [x] I have added to the [changelog](../blob/master/CHANGELOG.md#Unreleased)
- [x] I wrote/updated tests for new/changed code
- [ ] I removed all sensitive data **before every commit**, including API endpoints and keys, and internal links
- [x] I ran `fastlane ci` without errors
- [ ] All project file changes are replicated in `SampleSPM/SampleSPM.xcodeproj` project

Esakkiraja-Pothikannan committed:
feat: send device id in all RAT events (SDKCF-6625)

Esakkiraja-Pothikannan committed:
chore: update CHANGELOG.md

imaginaris commented on Sources/RInAppMessaging/Utilities/AnalyticsTracker.swift:
```diff
@@ -0,0 +1,16 @@
+internal struct AnalyticsTracker {
+    static func sendEventName(_ name: Constants.RAnalytics, dataObject: [String: Any]? = nil) {
+        var eventData = dataObject ?? [:]
+        eventData[Constants.RAnalytics.Keys.deviceID] = UIDevice.current.identifierForVendor?.uuidString
```
I think it would be better if device Id came from one source for the whole SDK code. To have a helper class which returns `UIDevice.current.identifierForVendor?.uuidString` as device ID and can be referenced here AND also in request classes

Esakkiraja-Pothikannan replied:
Done

imaginaris commented on Sources/RInAppMessaging/Utilities/AnalyticsTracker.swift:
```diff
@@ -0,0 +1,16 @@
+#endif
+
+internal struct AnalyticsTracker {
+    static func sendEventName(_ name: Constants.RAnalytics, dataObject: [String: Any]? = nil) {
```
Can't you just put `[:]` as default value instead of `nil`?
Then you could also remove line 11

Esakkiraja-Pothikannan replied:
Done

imaginaris reviewed (commented):
Isn't that more of improvement than feature?

Esakkiraja-Pothikannan committed:
chore: update review comments

- UIDevice extension created to fetch device id
- Refactored tracker function in AnalyticsTracker file

imaginaris commented on Sources/RInAppMessaging/Services/ImpressionService.swift:
```diff
@@ -53,7 +47,7 @@ internal class ImpressionService: ImpressionServiceType, HttpRequestable, TaskSc
         let impressionData = impressions.filter({ $0.type != .impression }).encodeForAnalytics()
-        AnalyticsBroadcaster.sendEventName(Constants.RAnalytics.impressionsEventName,
+        AnalyticsTracker.sendEventName(Constants.RAnalytics.impressionsEventName,
                                            dataObject: [Constants.RAnalytics.Keys.impressions: impressionData,
```
nit: please fix indentation in `dataObject` lines

Esakkiraja-Pothikannan replied:
Fixed

imaginaris commented on Sources/RInAppMessaging/Views/Presenters/BaseViewPresenter.swift:
```diff
@@ -86,7 +80,7 @@ extension BaseViewPresenter {
         let impressionData = [impression].encodeForAnalytics()
-        AnalyticsBroadcaster.sendEventName(Constants.RAnalytics.impressionsEventName,
+        AnalyticsTracker.sendEventName(Constants.RAnalytics.impressionsEventName,
                                            dataObject: [Constants.RAnalytics.Keys.impressions: impressionData,
```
nit: please fix indentation in `dataObject` lines

Esakkiraja-Pothikannan replied:
Fixed

imaginaris commented on Sources/RInAppMessaging/Views/Presenters/FullViewPresenter.swift:
```diff
@@ -155,7 +149,7 @@ internal class FullViewPresenter: BaseViewPresenter, FullViewPresenterType, Erro
     private func trackPushPrimerAction(didOptIn: Bool) {
-        AnalyticsBroadcaster.sendEventName(Constants.RAnalytics.pushPrimerEventName,
+        AnalyticsTracker.sendEventName(Constants.RAnalytics.pushPrimerEventName,
                                            dataObject: [Constants.RAnalytics.Keys.pushPermission: didOptIn ? 1 : 0,
```
nit: please fix indentation in `dataObject` lines

Esakkiraja-Pothikannan replied:
This is something should learn from you

Esakkiraja-Pothikannan replied:
Fixed

imaginaris reviewed (approved):
Left comments to fix the indentation but besides that - ✅

Esakkiraja-Pothikannan committed:
chore: fix indentation issue (SDKCF-6625)
